Analytics cleanup and new boottrack metric

ramcq opened the pull request:
Refactor/clean up the events we report to Google Analytics, making them more structured and useful. Adds numeric values for USB/disk selections, separate fields for version, source and personality of the OS, and a new boot track size metric.

https://phabricator.endlessm.com/T13644

ramcq committed:
record selected installation type with "Selected" event

Add a Dlg::SetSelectedInstallMethod to set the method and report any changes.

ramcq committed:
record USB size and storage size as numeric values

The label is kind of useless in analytics...

ramcq committed:
record metric with boot track size

ramcq committed:
add InstallerMethodToStr

ramcq committed:
add Dlg::TrackEvent wrapper which sets the section to the install method

ramcq committed:
record selected image version, personality and source as separate events

Far more useful for reporting than reporting the name.

ramcq committed:
refactor OnSelectFileNextClicked function

Avoid parsing the filename a 2nd time to get the version and personality - we
know these already from the local or remote file entry, and the version from
the remote entry is more accurate.

ramcq committed:
add Debug registry key to set analytics to validation server

wjt commented on src/endless/Analytics.cpp:
```diff
@@ -31,7 +32,8 @@ static UINT threadSendRequest(LPVOID pParam)
 			CString headers = _T("Content-type: application/x-www-form-urlencoded");
 			CHttpConnection *conn = session.GetHttpConnection(_T(SERVER_NAME), (INTERNET_PORT)SERVER_PORT);
-			CHttpFile *file = conn->OpenRequest(CHttpConnection::HTTP_VERB_POST, _T("collect"));
+			CString path = Analytics::instance()->m_debug ? _T("debug/collect") : _T("collect");
```
I'm not wild about accessing the global `Analytics::instance()` from this thread. Pass the debug flag (or the path) as the `pParam` to the thread? This would avoid having to define `m_debug` and declare the thread function a `friend` and so on.

wjt commented on src/endless/EndlessUsbToolDlg.cpp:
```diff
@@ -1926,7 +1953,7 @@ HRESULT CEndlessUsbToolDlg::OnTryEndlessSelected(IHTMLElement* pElement)
     FUNCTION_ENTER;
 
-	Analytics::instance()->eventTracking(_T(ELEMENT_FIRST_PAGE), _T("USBType"), _T("Live"));
+	eventTracking(_T("Selected"));
```
Shouldn't this go below the assignment to `m_selectedInstallMethod` below?

wjt replied:
Maybe it would be cleaner to define `::SetSelectedInstallMethod(InstallMethod_t method) { m_selectedInstallMethod = InstallMethod_t::TryEndless; eventTracking(_T("Selected")); }` and use this consistently?

wjt commented on src/endless/EndlessUsbToolDlg.cpp:
```diff
@@ -1922,8 +1922,16 @@ HRESULT CEndlessUsbToolDlg::OnInstallDualBootClicked(IHTMLElement* pElement)
 
 	if (UpdateDualBootTexts()) {
 		m_selectedInstallMethod = InstallMethod_t::UninstallEndless;
+		eventTracking(_T("Selected"));
```
As you said on Slack, a Selected event for UninstallEndless should be logged from `endless-uninstaller.exe` too.

wjt replied:
Controversial suggestion: get rid of the standalone path, just hide the "advanced options" bit at the bottom of the screen when run as `endless-uninstaller.exe`.

wjt commented on src/endless/EndlessUsbToolDlg.cpp:
```diff
@@ -2634,6 +2641,9 @@ HRESULT CEndlessUsbToolDlg::OnFirstPagePreviousClicked(IHTMLElement* pElement)
 
 HRESULT CEndlessUsbToolDlg::OnCreateEndlessUSBStickClicked(IHTMLElement* pElement)
 {
+	// install method is already set when the Endless USB dialog is created
```
This seems kind of error-prone…

wjt commented on src/endless/EndlessUsbToolDlg.cpp:
```diff
@@ -5080,6 +5080,8 @@ bool CEndlessUsbToolDlg::WriteMBRAndSBRToWinDrive(CEndlessUsbToolDlg *dlg, const
 		}
 	}
 
+	dlg->eventTracking(_T("BootTrackSizeKB"), _T(""), minStartingOffset / 1024);
```
I think we should log this in bytes, not KB. Rationale: the size is a multiple of the sector size, which is 512.

wjt commented on src/endless/EndlessUsbToolDlg.cpp:
```diff
@@ -2692,29 +2692,29 @@ HRESULT CEndlessUsbToolDlg::OnSelectFileNextClicked(IHTMLElement* pElement)
         }
-        selectedImage = CSTRING_GET_LAST(selectedImage, '\\');
+
+		personality = localEntry->personality;
```
This will crash in the error path. I think rather than leaning on dereferencing NULL to crash the process, we should advance to the error page if `.Lookup()` returns false.

ramcq committed:
detect and handle error cases for invalid local/remote file selections

Move the code which sets up USB device detection lower in the function so we
don't start anything in the error cases.

ramcq committed:
show normal first screen when called as endless-uninstaller.exe

We switch the Install button for Uninstall either when invoked as
endless-uninstaller.exe, or if C:\endless\endless.img exists. If the user
cancels the prompt we return to the main screen. This changes the analytics we
send a little, but the key events are unchanged, with the benefit of only
logging screen:UninstallPage and event:Uninstall,Selected if the user says yes
to the prompt. The UninstallFinished pseudo-screen has been removed, when the
session ends during an uninstall it will be the LastPage pseudo-screen.

ramcq committed:
rename InstallMethod_t consistently

LiveUsb,
ReformatterUsb,
CombinedUsb,
InstallDualBoot,
UninstallDualBoot

ramcq committed:
rename first page to advanced page

ramcq committed:
sed fest - rename identifiers to consistently use {Live,Reformatter,Combined}Usb

ramcq committed:
rename DualBootInstallButon to DualBootInstallButton

ramcq committed:
iconv -f UTF16 -t UTF8

UTF16 files don't work properly in git, grep, sed, etc...

ramcq committed:
after a verification failure, go back to the start rather than the advanced page

ramcq committed:
add missing break; in exception reporting to distinguish fatal / not

ramcq committed:
record Started, Completed, Cancelled, Failed and Closed events

Event reporting/analysis is far richer than exceptions, this will allow us to
split by demographic and see the flows of cancelling, errors, etc.

ramcq committed:
canceled -> cancelled

ramcq committed:
send new "Failed" event if uninstall fails too

ramcq committed:
remove unused magnifying glass resource

wjt commented on src/endless/EndlessUsbToolDlg.cpp:
```diff
@@ -1740,8 +1740,10 @@ void CEndlessUsbToolDlg::ErrorOccured(ErrorCause_t errorCause)
 	case ErrorCause_t::ErrorCauseVerificationFailed:
 	case ErrorCause_t::ErrorCauseWriteFailed:
 		fatal = FALSE;
+		break;
```
👏👏👏👏👏

wjt commented on src/endless/EndlessUsbToolDlg.cpp:
```diff
@@ -6030,8 +6030,10 @@ void CEndlessUsbToolDlg::QueryAndDoUninstall()
 	ShowWindow(SW_HIDE);
-	if (!UninstallDualBoot(this))
+	if (!UninstallDualBoot(this)) {
+		TrackEvent(_T("Failed"), ErrorCauseToStr(m_lastErrorCause));
```
Also log a "Started" event above?

ramcq committed:
record Started and Completed events for uninstall
